edx_prefectutils/hubspot_leads.py

```diff
@@ -1,68 +1,151 @@
+    sf_credentials: SFCredentials,
     sf_role: str,
     sf_database: str,
     sf_schema: str,
     braze_api_key: str,
     braze_api_server: str,
 ):
-    logger = prefect.context.get("logger")
+    logger = prefect.context.get('logger')
+    columns = [
+        'audience',
+        'braze_hs_contact',
+        'braze_programs',
+        'city',
+        'country_iso_code',
+        'createdate',
+        'email',
+        'first_conversion_event_name',
+        'firstname',
+        'highest_level_of_education',
+        'hs_object_id',
+        'lastname',
+        'lastmodifieddate',
+        'lead_source',
+        'lead_source_deal_value',
+        'lms_user_id',
+        'masters_leads',
+        'phone',
+        'utm_campaign',
+        'utm_medium',
+        'utm_source',
+    ]
 
-    sf_connection = snowflake.create_snowflake_connection(
-        sf_credentials,
-        sf_role,
-    )
-    sf_table = 'hubspot_leads'
-    query = """
-    SELECT lms_user_id, masters_leads FROM {table} WHERE lms_user_id IS NOT NULL
-    """.format(
-        table=snowflake.qualified_table_name(sf_database, sf_schema, sf_table),
-    )
-    cursor = sf_connection.cursor()
-    cursor.execute(query)
-    count = 0
-    batch = cursor.fetchmany(ATTRIBUTE_BATCH_SIZE)
-    while len(batch) > 0:
-        _send_hubspot_leads_to_braze_for_registered_users(
-            logger,
-            braze_api_server,
-            braze_api_key,
-            # Leads are separated by semicolons in the snowflake database
-            [LeadsData(row[0], row[1].split(';')) for row in batch],
-        )
-        count += len(batch)
-        if count % (ATTRIBUTE_BATCH_SIZE * 10) == 0:
-            logger.info("Updated %d user attributes in Braze", count)
-        batch = cursor.fetchmany(ATTRIBUTE_BATCH_SIZE)
-    if count > 0:
-        logger.info("Updated %d user attributes in Braze", count)
+    # There are two types of users:
+    # - Hubspot-alias-only users that don't have an account on the LMS
+    # - Users that have an lms_user_id (maybe never had hubspot, maybe did - 'identifying' aliases is done by the
+    #   platform-plugin-braze plugin that calls Braze's 'users/identify' API as users register on the LMS)
+    #
+    # We only want to set the "basic" fields like name/phone/email for alias-only users. Registered users get those
+    # fields from the LMS. But both categories get all the hubspot-specific fields.
+
+    alias_only_where = 'lms_user_id IS NULL'
+    registered_where = 'lms_user_id IS NOT NULL'
+
+    # Also, to reduce churn and data usage, only sync items that have changed since our last successful run
+    LAST_SUCCESS_KEY = 'hubspot-leads-last-success'
+    try:
+        last_success = get_key_value(key=LAST_SUCCESS_KEY)
+        alias_only_where += f" AND lastmodifieddate > '{last_success}'"
+        registered_where += f" AND lastmodifieddate > '{last_success}'"
+    except ValueError:
+        pass
```

Comment:
This is a new thing! Will this work alright? The docs suggests that we are limited on key/value pairings. But I don't see this being used elsewhere...

https://docs.prefect.io/orchestration/concepts/kv_store.html

(moving this comment to track with new version of code)

Reply:
Follow up - it will work. Tested and the values show up in the prefect KV store: https://cloud.prefect.io/team/kv